### PR TITLE
Make certbot constraint file independent from certbot-auto + update cryptography

### LINKED ISSUE #8648: Upgrade the pinned version of cryptography
In both `letsencrypt-auto-source/pieces/dependency-requirements.txt` and `tests/letstest/requirements.txt`, we pin `cryptography==2.8`. Over the weekend, `cryptography` did a release to fix a CVE. See https://github.com/pyca/cryptography/blob/b20507ae687dc2bf1841e7d7bc89cd4237177ae1/CHANGELOG.rst#332---2021-02-07. I doubt it significantly affects us, but I think we should just upgrade to be safe.

The minimum version we need is `cryptography>=3.3.2` which is probably the easiest thing to upgrade to here. Upgrading to an even newer version would be nice, however, it will likely require fixing snap/docker builds because `cryptography>=3.4.0` requires Rust to build its wheels. Making these builds work is something we have to do at some point, but whoever takes this may not want to do this now.

One final thing to flag here is I do not think we should be changing the version of `cryptography` pinned in `certbot-auto`. It shouldn't actually matter since `certbot-auto` refuses to install Certbot, but I don't think we should be making unnecessary changes to certbot-auto. Decoupling `letsencrypt-auto-source/pieces/dependency-requirements.txt` from certbot-auto and/or the rest of our pinning system is something that will need to be done at least partially here.

### PR DESCRIPTION
**Pull request by adferrand**
Fixes #8648 

This PR creates the file `tools/certbot_constraints.txt` that replaces `letsencrypt-auto-source/pieces/dependency-requirements.txt` to pin certbot dependencies when building Docker images, Snaps and the Windows installer.

In the process the `letsencrypt-auto-source/rebuild_dependencies.py` script that was used to construct the constraint file, is migrated to `tools/certbot_constraints.txt`.

I updated its content to:
* remove the authoritative constraints that were associated to certbot-auto or old Python versions
* pin cryptography to `3.3.2`
* use the bootstrap script from farm tests instead of `certbot-auto` to install the OS packages
* use Docker images whose OSes correspond to the OSs in farm tests

Finally, a new constraint file is generated from this script and it includes the upgrade for `cryptography`.

### CODE REVIEW

**Comment by adferrand** [comment]
For the record, the error encountered in letstest is due to the fact that `cryptography` 3.2 dropped support for OpenSSL 1.0.2, which is the version used by CentOS 7 / RHEL 7, and `test_sdists` tries to install Certbot dependencies without using binaries, which require a build of `cryptography` that consequently fails for these platforms.

**Comment by adferrand** [comment]
In this PR I trick pynsist to accept the cryptography wheel under Python 3.8 by renaming the wheel. Indeed the interpreter + ABI flag is `cp36-abi3` which reads as a non-universal wheel compatible which every Python 3.6+. But pynsist does not know that and consider that since the interpreter flag is `cp36`, then the wheel is compatible with Python 3.6 only.

The related PR on pynsist to properly fix the interpretation of wheel compatibility built under abi3 constraints is here: https://github.com/takluyver/pynsist/pull/222

File: tools/pip_install.py [hunk 1]
```diff
@@ -134,6 +133,7 @@ def main(args):
                 pip_install_with_print('--force-reinstall --no-deps --requirement "{0}"'
                                        .format(requirements))
 
+            print(' '.join(args))
```
Comment (bmw) [inline tools/pip_install.py:136]: Did you mean to add this print statement?

File: windows-installer/construct.py [hunk 2]
```diff
@@ -52,6 +53,21 @@ def _compile_wheels(repo_path, build_path, venv_python):
...
+    # with Python 3.6+). While technically valid to describe a wheel compliant with the Stable
+    # Application Binary Interface, this naming convention makes pynsist falsely think that the
+    # wheel is compatible with Python 3.6 only.
+    # Let's trick pynsist by renaming the wheel until this is fixed upstream.
+    for file in os.listdir(wheels_path):
+        # Given that our Python version is 3.8, this rename files like
+        # cryptography-VERSION-cpXX-abi3-win32.whl into cryptography-VERSION-cp38-abi3-win32.whl
+        renamed = re.sub(r'^(.*)-cp\d+-abi3-(\w+)\.whl$', r'\1-cp{0}{1}-abi3-\2.whl'
+                         .format(PYTHON_VERSION[0], PYTHON_VERSION[1]), file)
+        print(renamed)
```
Comment (bmw) [inline windows-installer/construct.py:67]: Did you mean to add this print statement?

File: tools/rebuild_certbot_constraints.py [hunk 3]
```diff
@@ -26,52 +26,41 @@
...
 #   - `certbot plugins` is called to check we have a healthy environment
 #   - finally current set of dependencies is extracted out of the docker using pip freeze
 SCRIPT = r"""#!/bin/sh
-set -e
+set -ex
 
 cd /tmp/certbot
-letsencrypt-auto-source/letsencrypt-auto --install-only -n
-PYVER=`/opt/eff.org/certbot/venv/bin/python --version 2>&1 | cut -d" " -f 2 | cut -d. -f1,2 | sed 's/\.//'`
+tests/letstest/scripts/bootstrap_os_packages.sh
```
Comment (bmw) [inline tools/rebuild_certbot_constraints.py:59]: I think we may want to move this script from `tests/letstest` to `tools` to help clarify this script isn't just for the test farm tests. What do you think? If you disagree, feel free to leave it.

File: tools/docker/core/Dockerfile [hunk 4]
```diff
@@ -14,10 +14,6 @@ WORKDIR /opt/certbot
 
 # Copy certbot code
 COPY CHANGELOG.md README.rst src/
-# We keep the relative path to the requirements file the same because, as of
-# writing this, tools/pip_install.py is used in the Dockerfile for Certbot
-# plugins and this script expects to find the requirements file there.
-COPY letsencrypt-auto-source/pieces/dependency-requirements.txt letsencrypt-auto-source/pieces/
 COPY tools tools
 COPY acme src/acme
 COPY certbot src/certbot
```
Comment (bmw) [inline tools/docker/core/Dockerfile:19]: There's a reference to `dependency-requirements.txt` below.

File: tools/rebuild_certbot_constraints.py [hunk 5]
```diff
@@ -215,11 +205,11 @@ def _write_requirements(dest_file, requirements, conflicts):
...
+# This is the flattened list of pinned packages to build certbot deployable artifacts.
 # To generate this, do (with docker and package hashin installed):
 # ```
-# letsencrypt-auto-source/rebuild_dependencies.py \\
-#   letsencrypt-auto-source/pieces/dependency-requirements.txt
+# tools/rebuild_certbot_contraints.py \\
+#   tools/certbot_constraints.txt
 # ```
 # If you want to update a single dependency, run commands similar to these:
 # ```
```
Comment (bmw) [inline tools/rebuild_certbot_constraints.py:215]: I think we should fix the reference to `dependency-requirements.txt` further down in this comment and regenerate this file.

**Review by bmw — APPROVED** [review]
I have some suggestions/questions below, but at a high level this looks good and fixes https://github.com/certbot/certbot/issues/8648 which I want resolved for the release so I'm merging this.

**Pull request merged**